[Team-15][iOS] IssueList TableView Convert & DIContainer Enhancement

SangHwi-Back opened the pull request:
# 2022/06/29 Pull Request

## 🔨 관련 Issue

- SangHwi-Back/issue-tracker#5
- SangHwi-Back/issue-tracker#8
- SangHwi-Back/issue-tracker#9
- SangHwi-Back/issue-tracker#26
- SangHwi-Back/issue-tracker#31

## 👨🏻‍💻 작업 내용

1. DIContainer 메모리 문제 해결
  - 저번 리뷰에서 피드백 해주신 메모리 문제를 Closure를 이용해서 해결하려 시도해보았습니다.
2. Issue List 를 표현하는 뷰를 UICollectionView 에서 UITableView 로 변경하였습니다.
  - 이로 인해 Swipe 액션을 손쉽게 넣고 네트워크 연결 없이 대부분의 작업을 수행할 수 있게 되었습니다.
  - 다음 PR 까지 Network 를 붙여서 실제 작업이 수행될 수 있도록 작업할 예정입니다.
  - **주의!** 잘못하고 IssueListTableViewDataSource, IssueListTableViewDelegate 파일을 제외하지 않았습니다. 앞의 두 파일은 사용하지 않습니다! 다음번 PR때는 이런 점 주의하도록 하겠습니다.
3. 불필요한 파일 삭제
4. 원래는 이슈 데이터 관련 네트워크 작업, 레이블 뷰 작업이 진행중이었습니다. 이 부분까지 PR에 초안이라도 넣어볼까 고민하였지만 한번에 너무 많은 수정사항을 리뷰 요청드리기는 부담이라 위의 작업만 리뷰를 요청드리게 되었습니다.

## 💭 고민한 점

- DIContainer Value 값 관리를 어떻게 해야하나?
    - `문제점 파악`
    - DIContainer는 Shard라는 싱글톤에 값을 Dictionary형태로 가지고 있고, 그안에 추상타입을 따르는 `구체타입을 생성하고, regist`를 하고 있었습니다. 이에 따라 생성된 값은 사용한 후에도 계속 남아있는 메모리 상의 문제점을 야기했습니다.
    - `해결 방안` 
    - **1. 사용후 에 Dictionary에서 값을 제거한다.**
        - 가장 간단하고 직관적인 해결법이라 생각했으나, 다시 사용하고 싶을때에는 또 다시 생성 -> regist -> resolve의 Flow를 이어나가야함. 현재 regist는 appdelegate에서 한번만 이루어지고 있는데 다시 regist하려면, 생성하는 로직을 어딘가에 추가를 해줘야하고, 이는 생성로직을 분리하고자했던 애초의 취지와 맞지 않다고 생각했습니다.
    - **2. factory Closure를 사용한다.**
        - Dictionary안의 Value에 구체타입을 넣어주는 것이 아니라, 추상타입을 따르는 구체타입을 리턴하는 Closure를 등록했습니다. 이에 따라, Dictionary안에는 Closure로직 밖에 없기 때문에 메모리 문제를 해결할 수 있고 생성로직을 appDelegate에 그대로 모아둘 수 있었습니다.
         
- UICollectionView나 UITableView의 DataSource, Delegate 를 따로 커스텀 클래스로 만들어야 할까?
    - `개인적인 결론`
    - DataSource는 Model이나 ViewModel이 될 수 있을 거라 판단했습니다. UITableView의 Data에 반응하기 때문에, UITableView를 소유한 ViewController가 DataSource라는 Model/ViewModel과 소통하는 방식이 MVC/MVVM Pattern과 유사하다고 생각하였습니다.
    - 하지만, 저희가 MVVM 을 선택하면서 ViewModel 이 Data 를 점유하는 순간 DataSource의 역할이 적어졌습니다.
    - `질문드리고 싶은 사항`
    - MVVM에서 DataSource를 나누어야 하는 이유가 있을지 궁금합니다.
    - Delegate(UITableViewDelegate, UICollectionViewDelegate)의 경우도 커스텀 클래스로 나눈 후 ViewController가 소유하도록 해야 할까요? 어차피 ViewController, Model(혹은 ViewModel)과 상호작용해야 하기 때문에 별다른 의미는 없을 것이라 생각했습니다.

## 🗓 다음 PR까지 진행할 작업 목록

- **Beck**
    - Issue 관련 GitHub API를 연결할 UseCase, Model, Repository, Service를 연결할 것입니다.
    - 가능하다면 Filter 뷰까지 구현하여 필터링 되는 부분까지 보여드릴 수 있다면 좋겠네요!
- **PigBag**
    - 현재 Label을 보여주는 TableView를 구현중입니다.
    label의 색상을 조절하는 기능을 추가로 구현한 뒤에, IssueView와 연결할 예정입니다!

SangHwi-Back committed:
chore: resolve conflict collectionview

SangHwi-Back committed:
chore: conflict fix

SangHwi-Back committed:
chore: resolve conflict collectionview

SangHwi-Back committed:
feature: convert collection view to table view

SangHwi-Back committed:
feature: convert collection view to table view

SangHwi-Back committed:
chore: conflict fix

SangHwi-Back committed:
chore: add commit not contained

SangHwi-Back committed:
chore: add commit not contained

SangHwi-Back committed:
chore: resolve all commits

SangHwi-Back committed:
chore: resolve all commits

SangHwi-Back committed:
chore: code clean and add titleLabel

SangHwi-Back committed:
chore: code clean and add titleLabel

SangHwi-Back committed:
feature: cell swipe action resolve

SangHwi-Back committed:
feature: cell swipe action resolve

SangHwi-Back committed:
feature: devide bar button item function

SangHwi-Back committed:
feature: devide bar button item function

SangHwi-Back committed:
refactor: remove unused files

SangHwi-Back committed:
refactor: remove unused files

Commit:
Merge pull request #32 from SangHwi-Back/issueConvertTableView2

[Feature] Issue List Swipe and Bar button Action

SangHwi-Back committed:
Merge branch 'develop' of https://github.com/SangHwi-Back/issue-tracker into develop

piggyse committed:
[refactor] DIContainer 리팩토링

Contents
- DIContainer의 Dictionary Value값 구체타입에서 생성 Closure로 변경
- 실제 구현부 클로저 정의 및 사용으로 리팩토링

Todo
- 바뀐 프로토콜 메서드 다른 Container에도 적용.

SangHwi-Back committed:
Merge pull request #34 from SangHwi-Back/refactor/Container

Refactor/container

piggyse committed:
[refactor] UseCseContainer

Contents
- Resolvable 프로토콜 수정
- 필요없는 파일 제거

Commit:
Merge branch 'develop' into refactor/Container

SangHwi-Back committed:
Merge pull request #35 from SangHwi-Back/refactor/Container

[refactor] UseCseContainer

piggyse commented on IssueTracker15/IssueTracker15/UseCaseInterface/UseCaseContainer.swift:
```diff
@@ -7,23 +7,23 @@
-        let identifier = ObjectIdentifier(T.self)
-        guard let useCase = instance as? Value else { return }
-        self.useCases[identifier] = useCase
+    mutating func regist<T: UseCaseResponsible>(type: T.Type, make: @escaping Creator) {
```
Resolvable 프로토콜에서 extension을 활용해 Default regist, resolve함수를 제공하고 있지만, 제네릭 T가 UseCaseResponsible을 따르는 것들만 인자로 들어올 수 있게 하고싶어서 재정의 했습니다.

SangHwi-Back commented on IssueTracker15/IssueTracker15/Issue/IssueList/View/IssueListTableViewDelegate.swift:
```diff
@@ -0,0 +1,12 @@
+
+import UIKit
+
+class IssueListTableViewDelegate: NSObject, UITableViewDelegate {
```
이 파일(클래스)은 사용하지 않습니다. 다음 PR 에는 없앨 예정이오니 참고 바랍니다.

SangHwi-Back commented on IssueTracker15/IssueTracker15/Issue/IssueList/View/IssueListTableViewDataSource.swift:
```diff
@@ -0,0 +1,59 @@
+
+import UIKit
+
+class IssueListTableViewDataSource: NSObject, UITableViewDataSource, ViewBindable {
```
이 파일(클래스)은 사용하지 않습니다. 다음 PR 에는 없앨 예정이오니 참고 바랍니다.

SangHwi-Back committed:
feature: Add Issue List/Modify UseCase, Model, Repository, Service

SangHwi-Back committed:
feature: decoding error fix

- Issue List Success

piggyse committed:
[feature] Setting LabelView

Contents
- SetNavigation Title
- create TableViewController for LabelVC

piggyse committed:
[feature] Add Label Button

Contents
- Create Custom Add Label Navigation bar Button
- adapt to VC

SangHwi-Back committed:
feature: Add Service container in Repository

piggyse committed:
[feature] Add Labels in TableViewCell

Contents
- IssueLabel과 설명 Label Cell에 추가
- TableViewController -> ViewController로 변경

piggyse committed:
[feature] Add IssueLabel Struct & ReName

Contents
- IssuLabel Struct 생성
- IssueLabelButton -> Title 로 변경

Commit:
Merge pull request #37 from SangHwi-Back/issueRequestModel

Issue request model

SangHwi-Back committed:
Merge pull request #38 from SangHwi-Back/feature/issueLabel

Feature/ Issue label

piggyse committed:
[refactor] ViewBindable

Contents
- setVC Extension으로 빼기.

SangHwi-Back committed:
refactor: UseCaseResponsible param 파라미터 추가

piggyse committed:
[fix] TableView Cell Height Error

SangHwi-Back committed:
feature: Repository adds specific service

- 세션이 해제되는 문제가 생겨 확인해보니 컨테이너 문제가 있음.
- 관련 이슈에 대해 차차 해결해 나갈 예정.

piggyse committed:
[refactor] IssueTitle Button

Contents
- 상수값으로 CornerRadius 주던 것 리팩토링

SangHwi-Back committed:
feature: Container, Wrapper 구조 일부 변경

Commit:
Merge pull request #39 from SangHwi-Back/networkEnhancement

[Feature] Network Container Enhancement

Commit:
Merge pull request #40 from SangHwi-Back/feature/LabelView

Feature/label view

SangHwi-Back commented on IssueTracker15/IssueTracker15/Repository/Repository.swift:
```diff
@@ -2,27 +2,26 @@
-    
-    func getNetworkService<T: Codable>(resultType: T.Type) -> NetworkService<T>? {
-        let service = NetworkService<T>()
+    func getRESTNetworkService<T: Codable>(type: T.Type) -> NetworkService<T>? {
```
serviceWrapper는 Service 를 생성하는 클로저를 저장해두는 객체로서 사용합니다. 하지만, Session Deinitialized 때문에 객체를 전달하는 방식으로 변경하였습니다.

SangHwi-Back commented on IssueTracker15/IssueTracker15/Repository/Repository.swift:
```diff
@@ -2,27 +2,26 @@
+        
+        guard let service = serviceWrapper.resolve(type: type.self) else {
+            
+            serviceWrapper.regist(type: type.self, make: NetworkService<T>())
```
원래 이 부분은 NetworkService<T>() 가 아닌 { NetworkService<T>() } 였습니다. 클로저를 전달하는 것이지요.

근데 왼쪽처럼 하게 되면 Session Deinitialized 가 안되고, 오른쪽처럼 하면 Session Deinitialized가 됩니다. 그리고 오른쪽이 저희가 의도한 대로 나중에 resolve 함수를 이용해 의도한 서비스를 가져오고 있습니다.

이전에 Container로 객체를 주입하면 메모리 문제가 있을 수 있다 해주신 기억이 있습니다. 지금의 방법은 어떨까요? 결국 마지막 regist를 구현한 Resolvable 프로토콜 에서는 객체를 반환하는 클로저가 됩니다.

제 생각엔 이미 점유한 메모리의 객체를 참조하여 전달할 뿐인 것 같아 불안한 느낌입니다.

dev-Lena replied:
Repository 클래스가 메모리에서 해제되면 이 부분도 자연스럽게 해제될 것이라고 예상이 되는데요.
혹시 명시적으로 메모리에서 해제하여 메모리 관리를 하고 싶다면,
[RxSwift의 DisposBag의 구현부](https://github.com/ReactiveX/RxSwift/blob/main/RxSwift/Disposables/DisposeBag.swift)를 한번 보시면 좋을 것 같습니다.
이에 대해서 RxSwift의 다른 부분은 볼 필요는 없고, `DisposeBag.swift:63` 에 있는 dispose() 메서드를 참고하면 좋을 것 같은데요. 제가 말씀드리고 싶은 것은 이런 아이디어를 참고해 
더이상 필요가 없다고 확신하는 상황에서 deinit 에서나 레퍼런스를 관리하는(혹은 관리를 담당하는게 자연스러운) 객체에서 안전하게 메모리 할당 해주는 방법 한 번 생각해보시면 좋을 것 같습니다( 추가 구현이 필요하다는 부분은 아닙니다!)

SangHwi-Back commented on IssueTracker15/IssueTracker15/Container/Resolvable.swift:
```diff
@@ -7,7 +7,20 @@
+        let identifier = ObjectIdentifier(type)
+        self.map[identifier] = {
+            make
+        }
```
원래는 Repository 혹은 호출부에서 클로저를 전달했지만 메모리 해제가 일어나서 Value를 그대로 반환하는 클로저가 되었습니다.

이 방식은 메모리 측면에서, 이전에 객체 자체를 주입하는 방식과 별 차이가 없는걸까요? 의견 부탁드립니다. 🙏🙏🙏

dev-Lena replied:
이전에 객체 자체를 주입하는 방식이라고 하면 이 [커밋](https://github.com/SangHwi-Back/issue-tracker/commit/17e6eaf64be170b7b01a86354fd215463e32bcb3) 말씀하시는거 맞나요?
변경 사항이 escaping 클로저로 호출부에서 넣어주던 것을 파라미터로 전달하여 그대로 넣어주는 걸로 변경된게 맞다면, 
escaping 클로저가 호출부의 객체에서 메모리에 잡혀있는 부분만 제외하면 크게 다른 점은 없는 것 같습니다~!

piggyse commented on IssueTracker15/IssueTracker15/Label/View/IssueTitleButton.swift:
```diff
@@ -0,0 +1,37 @@
+        super.layoutSubviews()
+        self.layer.cornerRadius = self.bounds.height / 2
+        self.clipsToBounds = true
+    }
```
init될 당시에는 bound값이 0이므로 self.bounds.height / 2 을 사용할 수 없었습니다.
따라서 LayoutSubViews에서 CornerRadius값을 바꾸어 보았습니다.

질문 🙋‍♂️
layoutSubviews에서 Configuartion값을 바꾸려고 시도하면, 앱이 무한루프에 들어간듯이 앱 자체가 실행이 되지 않았습니다.
~~~
self.configuration?.background.cornerRadius = self.bounds.height / 2
~~~
layoutSubView에서 뷰의 속성을 바꾸면 무한루프에 도는 것일까요?? ㅠㅠ

dev-Lena replied:
`layoutSubviews() `내부에서 위 코드를 호출해서 문제가 됐다기 보다는 iOS 15에 추가된 UIButton의 configuration을 지정하는 방법이 잘못되진 않았을까 싶은데, 
아래 두 문서를 참고해서 UIButton의 Configuration의 cornerRadius를 바꿔보시는건 어떨까요?
* [UIButton.Configuration.CornerStyle](https://developer.apple.com/documentation/uikit/uibutton/configuration/cornerstyle)
* [how to mark custom button style with uibuttonconfiguration](https://sarunw.com/posts/how-to-mark-custom-button-style-with-uibuttonconfiguration/)

dev-Lena reviewed (approved):
Beck과 pigBag 안녕하세요! 이번 PR도 많은 고민을 하신 것 같네요! 수고하셨습니다 🧚🏻‍♂️
특히 네트워크 객체의 메모리에 대해서 어떻게 관리할지에 대해서도 고민을 하신 것 같은데요. 이 시점에서 네트워크 관련 스트림 개념과 그 스트림 안에서의 메모리 관리는 어떻게 하면 좋을지 고민해보시면 좋을 것 같습니다 👍

아래는 디스크립션에 남겨주신 생각들에 대한 저의 의견(질문)과 답변입니다! 🤓

> ### 고민한 점 - DIContainer Value 값 관리를 어떻게 해야하나?
> 2. factory Closure를 사용한다.
> Dictionary안에는 Closure로직 밖에 없기 때문에 메모리 문제를 해결할 수 있고

🧐 클로저가 Dictionary에 들어있으면 메모리에 살아있는게 아닌가 생각이 드는데, Closure 로직이 있어 메모리 문제를 해결할 수 있다는게 어떤 의미인지 좀 자세히 설명해주실 수 있을까요?

> ❓UICollectionView나 UITableView의 DataSource, Delegate 를 따로 커스텀 클래스로 만들어야 할까?
> MVVM에서 DataSource를 나누어야 하는 이유가 있을지 궁금합니다.

사실 DataSource를 무엇으로 볼지에 대해서는 의견이 분분한 면이 없지 않은 것 같습니다. 
Beck과 pigBag도 디스크립션에 아래와 같이 의견을 적어주셨는데요.
> DataSource는 Model이나 ViewModel이 될 수 있을 거라 판단했습니다. 
> 하지만, 저희가 MVVM 을 선택하면서 ViewModel 이 Data 를 점유하는 순간 DataSource의 역할이 적어졌습니다.

이 부분을 보면서 저는 🧐DataSource의 역할이 커야 하는가? 라는 의문이 먼저 생겼습니다. 그리고 개인적으로는 DataSource는 View에 보여질 Data를 넣어주는 곳이지만 비즈니스 로직을 담당하는 역할을 하는 객체는 아니라고 생각합니다. 그 점에서는 DataSource는 ViewModel보다 View에 가깝다고 생각하고 있고, 그렇기 때문에 DataSource는 ViewModel에서 비즈니스 로직이 처리된 데이터를 전달받아 View에 전달하는 역할로도 충분하다고 생각합니다.

> ❓Delegate(UITableViewDelegate, UICollectionViewDelegate)의 경우도 커스텀 클래스로 나눈 후 ViewController가 소유하도록 해야 할까요? 어차피 ViewController, Model(혹은 ViewModel)과 상호작용해야 하기 때문에 별다른 의미는 없을 것이라 생각했습니다.

말씀하신대로 Delegate 에서는 VC 혹은 VC가 소유하고 있는 객체들과 상호작용하는 하는 경우가 많기 때문에 나누는게 오히려 번거로운 일이 될 수 있다고 생각합니다.

SangHwi-Back commented:
@dev-Lena 리뷰 감사드립니다. RxSwift 는 개인적으로도 공부하고 있는 라이브러리기 때문에 Repository 를 고쳐보는 건 재미있는 일이 될 것 같네요.

3주 동안 고생 많으셨습니다!
